### DISCUSSION TITLE
Image extraction

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#2582] [Looking for help]
There is one image in the PDF but page.get_images() gives the following output
[(18, 0, 354, 118, 8, 'DeviceRGB', '', 'I0', 'DCTDecode'), (19, 0, 1181, 1772, 8, 'DeviceRGB', '', 'I1', 'DCTDecode'), (18, 0, 354, 118, 8, 'DeviceRGB', '', 'I0', 'DCTDecode')]

Can you let me know how the image file and bbox coordinates can be extracted?

In the internal structure it is represented as Form XObject with  /BBox [ 0 0 595.276 76.53543 ] but the bounding box  is not getting drawn with these values.
[example_010.pdf](https://github.com/pymupdf/PyMuPDF/files/12272759/example_010.pdf)

**Comment by JorjMcKie** [comment]
The method `page.get_images()` reports what is contained in the PDF page object definition. This is not necessarily a list of images actually displayed by the page. Please read the documentation about that method and section **"Block Dictionaries"** within the `TextPage` chapter.
Page 1 of your example **_displays_** only one image. Execute first `page.clean_contents()` and **then** `page.get_images()`, and you will see only one item. This is because cleaning synchronizes the object definition with the appearance source of the page.
Otherwise, detecting the bbox of an image works via `page.get_images_rects(xref)`.

**Comment by meghanaviyyapu** [comment]
Thanks. May I know how page.get_images_rects(xref) computes the bounding box values?

**Reply by JorjMcKie** [reply]
This happens based on `page.get_image_info()` combined with `page.get_images` - see the source in `utils.py`.

**Reply by meghanaviyyapu** [reply]
Is it possible to calculate bbox values with width and height of image, page coordinates?

**Reply by JorjMcKie** [reply]
> Is it possible to calculate bbox values with width and height of image, page coordinates?

What does that mean? If you are asking for the relationship between image dimension (width, height) and the bbox dimension on the page, then please look at the documentation [here](https://pymupdf.readthedocs.io/en/latest/app3.html#imagetransformation).

**Comment by meghanaviyyapu** [comment]
Had a look.
Would like to know the internal working of "page.get_image_bbox("fzImg0", transform=True)" or how the bbox values can be calculated without using page.get_image_bbox()

**Reply by JorjMcKie** [reply]
Well, there are the source codes in `utils.py` and `fitz.i`. The main logic is in `get_image_info(xrefs=True)` which uses the page's `TextPage`. Using the MD5 code of the extracted image, a match is tried with the items in `page.get_images()`.

**Comment by meghanaviyyapu** [comment]
Thanks for the explanation
